### PR TITLE
platforms proposal: make installation configurable at the suite level

### PR DESCRIPTION
**Pull request by dpmatthews**
Also fix a mistake in the config changes proposal.

### CODE REVIEW

File: docs/proposal-platforms.md [hunk 1]
```diff
@@ -277,10 +278,33 @@ suite-run functionality:
...
+         use of install tasks as a workaround.
+       * Proposal:
+         * Install the same items to all install targets.
+         * Allow the installation items to be configured at the suite level via
+           ones of the following ways:
+           * A new setting `[scheduler]install` with a default value of
+             `app bin etc lib` (the items in `.service` are always installed).
+             Note that if we want to be able to support copying top level files
+             as well as directories then we would need to do something like
+             `app/ bin/ etc/ lib/ extra-dir-to-copy/ extra-file-to-copy`.
```
Comment (oliver-sanders) [inline docs/proposal-platforms.md:297]: Note that `[scheduler]install` should not accept patterns or anything other than top-level files or directories.

Comment (hjoliver) [reply docs/proposal-platforms.md:297]: Why is that exactly?

**Comment by dpmatthews** [comment]
@hjoliver - @datamel is working on this now so we need to at least decide what to implement initially.
I'd be happy to either support both or possibly just support the rsync filter initially (which is trivial to implement) and then add the config item later once we've had a chance to try out the functionality (if we decide we need it).

File: docs/proposal-platforms.md (hunk 1 above)
Comment (oliver-sanders) [reply docs/proposal-platforms.md:297]: 1) Force adoption of standard permitted directories (`bin`, `lib`, `app`).
2) Prevent overly generic globs (e.g. `*`).
3) Multi-level includes would require extra logic in forming the `rsync` command.

**Comment by datamel** [comment]
Can I confirm that the intention is, when the `[scheduler]install` setting is present in a suite.rc, the default values of `app bin etc lib` will all be overridden? (and that the .service is essentially hard-coded / non-configurable).

I ask because:

> Note that if we want to be able to support copying top level files as well as directories then we would need to do something like app/ bin/ etc/ lib/ extra-dir-to-copy/ extra-file-to-copy.

suggests that we would need to allow additional files and directories to be appended i.e. would we then expect the user to 'spell out' those defaults themselves to be able to add additional sources? i.e. in what scenario will these default 'abel' :) folders be overridden; the moment a new source is added in `[scheduler]install`?

I agree with @oliver-sanders that we can only really offer top-level sources to be added in the suite.rc. Any depeer would presume some rsync knowledge, which wouldn't be intuitive in a non-rsync config file (and to all users) e.g. to add a file deep in you also need to add the directories it is under separately, also.

I like the idea of supporting both. They play well together although the order on command line is important. We would typically keep the `exclude="*"` at the end but putting the filter before includes would mean filters take prescedence. Again, we need to clarify how this sits then, with the 'abel' folders, which will be coming in via `--include`.

We just need to be clear about what 'trumps' what between defaults, suite.rc and the filters. Here in the proposal and then in the documentation.

Useful to note that `-F` can be used in place of `--filter=': .rsync-filter'` and `-FF` will also exclude the .rsync-filter itself from being copied.

**Comment by dpmatthews** [comment]
> Can I confirm that the intention is, when the `[scheduler]install` setting is present in a suite.rc, the default values of `app bin etc lib` will all be overridden? (and that the .service is essentially hard-coded / non-configurable).

I was suggesting that just in case anyone wanted to remove some of the defaults but that's not needed if we support the rsync filter as well.

> I like the idea of supporting both. They play well together although the order on command line is important. We would typically keep the `exclude="*"` at the end but putting the filter before includes would mean filters take prescedence. Again, we need to clarify how this sits then, with the 'abel' folders, which will be coming in via `--include`.

I think the filter should take precedence (so that the users can override the defaults if they want to). I'll update the proposal.

File: docs/proposal-platforms.md (hunk 1 above)
Comment (dpmatthews) [reply docs/proposal-platforms.md]: I can't see any reason not to allow patterns so long as we explicitly exclude anything which should never be copied. This makes it easy for users to mimic the current `rose suite-run behaviour`. See updated proposal.

Comment (dpmatthews) [reply docs/proposal-platforms.md]: Resolving this - please comment on the updated proposal

File: docs/proposal-platforms.md [hunk 2]
```diff
@@ -286,32 +286,44 @@ suite-run functionality:
...
-         sure if this is deliberate) - exclude would be very bad in these cases.
-         Therefore, include is the safer option.
+         * The rsync filter options will be applied in the following order (note
+           that the first matching pattern is acted on):
+           * Include files required by cylc:
+             `--include='/.service/' --include='/.service/server.key'`
+           * Exclude directories which should never be copied:
+             `--exclude='.service/***' --exclude='log' --exclude='share' --exclude='work'`
+           * Apply any rules define in an `.rsync-filter` file:
+             `--filter=': .rsync-filter'`
```
Comment (datamel) [inline docs/proposal-platforms.md:320]: Would you like users to specify in their suite design that they are using a filter file, or do you just want Cylc to check and use the filter file if one happens to exist?

Comment (dpmatthews) [reply docs/proposal-platforms.md:320]: I'm suggesting that cylc always uses this rsync option for the installation and then rsync will use the file if it exists.

File: docs/proposal-platforms.md [hunk 3]
```diff
@@ -277,17 +278,52 @@ suite-run functionality:
...
-         sure if this is deliberate) - exclude would be very bad in these cases.
-         Therefore, include is the safer option.
+       * If we allow global configuration this is likely to lead to portability
+         issues where a suite only works if you modify your global config. If
+         additional items need to be installed it's better to force these to be
+         configured in the suite. Note that this will require changes to some
+         Cylc 7 suites in order for them to work at Cylc 8.
+       * Hopefully it should be fine to install the same files to all install
+         targets. If there are cases where this is an issue we can recommend the
+         use of install tasks as a workaround.
```
Comment (hjoliver) [inline docs/proposal-platforms.md:288]: Not entirely sure about this, but its fine as a first cut.

**Comment by dpmatthews** [comment]
All feedback addressed so merging this. We can fine tune if necessary as part of the implementation.

**Pull request merged**